### PR TITLE
feat(client): ts-migrate full /client/utils directory

### PR DESCRIPTION
**Pull request by samarmohan**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Everything in the `client/utils` folder was converted!

Refers #42256.

### CODE REVIEW

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
Hey there,

Thank you, for taking the time to contribute. Could you please change all files that do not use JSX from `.tsx` to `.ts`?

**Comment by samarmohan** [comment]
> Hey there,
> 
> Thank you, for taking the time to contribute. Could you please change all files that do not use JSX from `.tsx` to `.ts`?

Done

**Comment by samarmohan** [comment]
@gikf when will the "merge-party" be held?

**Comment by gikf** [comment]
> @gikf when will the "merge-party" be held?

It wasn't decided yet, per https://github.com/freeCodeCamp/freeCodeCamp/issues/42256#issuecomment-851673768. Details likely will be announced in that issue.

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
Hey again,

As the bot says above, no non-english files should be changed, as we have an automated workflow to rebase them with any changes to English ones, and then they get translated. Please revert the changes to the `i18n` docs.

**Review by raisedadead — COMMENTED** [review]
Hey there,

Thank you for taking the time to contribute to the TypeScript migration. 

The main issue has been updated a lot over the last few days, and so we suggest you re-read it and update this pull request to match if you have the time.

The main goals for a PR are:
- [ ] Changing the filename from `PascalCase` to `kebab-case` and ensuring that the imports in code are all still valid.
- [ ] Changing the file extension to .tsx (if a file has React JSX syntax)
- [ ] Fixing as many TypeScript and Eslint errors as you can while refactoring the code.

If you need any assistance, please reach out in our [contributing chat room](https://chat.freecodecamp.org/channel/contributors).

File: client/utils/build-challenges.ts
Comment (ShaunSHamilton) [inline client/utils/build-challenges.ts:9]: @ojeytonwilliams Was there not a reason this was being required instead of imported?

Comment (samarmohan) [reply client/utils/build-challenges.ts:9]: Sorry, I’ll fix that

File: client/utils/challenge-types.ts
Comment (ShaunSHamilton) [inline client/utils/challenge-types.ts:4]: I do not understand why these changes are here.

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
Hello there. Thank you, for taking the time to contribute.

Some of these changes should not be here, including the changes to any non-english files.

File: client/utils/build-challenges.ts
Comment (ShaunSHamilton) [reply client/utils/build-challenges.ts:9]: @samarmohan There may not be a need to fix this. I am getting clarification from Oliver, whether or not this change could break things.

File: client/utils/challenge-types.ts
Comment (samarmohan) [reply client/utils/challenge-types.ts:4]: They were being exported somewhere else, I changed it from a commonjs to an ESM export.

Comment (ShaunSHamilton) [reply client/utils/challenge-types.ts:4]: This is being exported on line 19, already.

File: client/utils/build-challenges.ts
Comment (ojeytonwilliams) [reply client/utils/build-challenges.ts:9]: No reason I can think of. It won't fail quietly, so if the build works at all with this change it should be fine.

File: client/utils/challenge-types.ts
Comment (samarmohan) [reply client/utils/challenge-types.ts:4]: I can't re-export a const so instead of doing it around line 19 I did it where it was initialized.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Hey @samarmohan, could you remove the i18n files, please?  

```
docs/i18n/chinese/how-to-work-on-coding-challenges.md
docs/i18n/espanol/how-to-work-on-coding-challenges.md
docs/i18n/portuguese/how-to-work-on-coding-challenges.md
```

Crowdin will create PRs for these files once they've gone through the translation workflow.

**Comment by samarmohan** [comment]
@moT01 thank you so so so much!

**Comment by moT01** [comment]
No problem @samarmohan - thanks for contributing to freeCodeCamp 🎉 

I am blocking this for now while we look into a separate issue.

**Review by raisedadead — CHANGES_REQUESTED** [review]
Thanks for the PR. 

I pulled this down, rebase and clean up the commits. 

It looks like this will need a lot of work before we can merge it. With the merge-party over, we are working on making the target branch of this PR stable.

We will not be able to review or merge this PR just yet; you are welcome to try and ensure that the lining and tests are passing if you are interested. If you do, we would also need you to remove all the lining bypasses.

Alternatively, you can leave the PR as is, and we will close it when someone gets around with a replacement PR.

If you have any questions or need help, please reach out to us in the contributors chat. Cheers.

**Pull request closed without merging**
